### ISSUE TITLE
GenMeshVertexNormals gens bad normals and also doesn't weight normal smoothing by poly area

### ISSUE THREAD

**Issue by alecazam** [assimp/assimp#3481]
I'm seeing many online reports about assimp flipping normals, and we have that case on a simple obj file.  It's a tree with a thin canopy of leaves that has two sides, but the verts are very close.  It looks like the normal gen is just looking for closeness of vertices and not actually walking the topology of the indices.  So the tree trunk is fine, but the canopy has incorrect normals.  

Another issue is that I looked at the GenSmoothNormals call, and it's simply walking around vertices and adding up normals without weighting them by the triangle areas.  It's a simple change to store the area, and then normalize for dot product with the limit case.  Or don't normalize after the cross-product if there isn't a limit.

This flag is included in the HighQuality settings, and flipping normals isn't great for our visuals.   Apple's 3D preview shows the obj correctly before going through assimp. 

This line below is where normals aren't weighted.  Another array is needed to store the weights, or for the non-limit case just don't normalize until after normals are combined and they will naturally be weighted.  Also watch out for zero area tris.  

const aiVector3D vNor = ((*pV2 - *pV1) ^ (*pV3 - *pV1)).NormalizeSafe();

**To Reproduce**
1. Run assimp with aiProcess_GenSmoothNormals

**Expected behavior**
I expect normals to be the same or close to the original obj after running smoothing.  Normals shouldn't flip esp. with HighQuality setting.

**Screenshots**
This is a shot of the original leaves.
![image](https://user-images.githubusercontent.com/3750359/97221379-057dfa80-178a-11eb-8bb0-42b21af278cb.png)

**Desktop (please complete the following information):**
 - OS: macOS 
 - Browser n/a
 - Version 10.15.7

**Additional context**
None

**Comment by alecazam** [comment]
It's like this call is trying to re-weld with a radius around the vertex to smooth normals and it's working too hard.  Generating vertex normals is simple assuming that you have generated indices.

1. Weld vertices and remap indices or use existing indices depending on flags
2. Each triangle of indices tell which vertices a face normal contributes to, and the area = length(cross(v1-v0, v2-v0)) says how much to weight.   This will fix thin geometry and also not weight normals too much by tiny sliver triangles, and it's fast O(n) time.

**Comment by alecazam** [comment]
Also looks like CalcProcessTangents is using similar code that has the same issue.  Smoothing tangents/bitangents should also no be rewelding vertices, and should walk the index list.

**Comment by Garux** [comment]
I'm experiencing flipped normals after gen/genSmooth funcs too, but considering that `aiProcess_FlipWindingOrder` is used, and gen funcs do not take this into account, it's logical.

**Comment by alecazam** [comment]
The normals in our model are fine, but the geometry is thin. The merging of thin features off pure positional proximity and not topology leads the problem.  Even if you took a thin box, then assimp would exhibit the errors that I encountered.

**Comment by tellypresence** [comment]
@alecazam Is this still a problem?

**Comment by alecazam** [comment]
Yes, I mean our pipeline is still using assimp.  But the above thin tree canopy geometry had to be thickened to avoid the welding.  This seems to just be a fundamental no-no to ignore topology in any geometry processing from what I've written.

**Comment by tellypresence** [comment]
@alecazam Sounds like there's a workaround, please reopen if this happens in other situations
